sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -210,6 +204,39 @@ class NodeHttpClient implements HttpClient {
     return new Promise<http.IncomingMessage>((resolve, reject) => {
       const req = isInsecure ? http.request(options) : https.request(options);
 
+      // The `connectionTimeoutInMs` property defines the number of milliseconds a request can take on stablishing a connection.
+      // If `connectionTimeoutInMs` is not set, the `timeout` property is used for the same purpose.
+      // Defaults to 0, which disables the timeout.
+      const connectionTimeoutInMs = request.connectionTimeoutInMs || request.timeout;
+      let connectionTimeout: NodeJS.Timeout;
+      if (connectionTimeoutInMs > 0) {
+        req.on("socket", (socket) => {
+          // If connecting is true, no connection has been established yet.
+          if (socket.connecting) {
+            connectionTimeout = setTimeout(() => {
+              socket.destroy();
+              abortController.abort();
+            }, connectionTimeoutInMs);
+          }
+          // If connecting is false, the socket is already connected.
+        });
+      }
+
+      // The `timeout` property defines the number of milliseconds a request can take before automatically being terminated.
+      // If the socket is connected before this timeout expires, we only wait for the remaining time before throwing.
+      // Defaults to 0, which disables the timeout.
+      if (request.timeout > 0) {
+        const connectingStartDate = Date.now();
+        req.on("connect", () => {
+          clearTimeout(connectionTimeout);
+          const connectingTimeElapsed = Date.now() - connectingStartDate;
```

Comment:
Why do all this extra work instead of just firing off the timeout like before? It's not like we can double-abort. Nor can we abort a request that already finished.